Improve contributing guide (#618)
percona/pmm-managed: CONTRIBUTING.md

# Contributing notes

pmm-managed is highly integrated with PMM Server environment. As a result, development and testing is done partially inside PMM Server container which we call "devcontainer". You can take a look at the architecture [here](https://www.percona.com/doc/percona-monitoring-and-management/2.x/details/architecture.html)

# Devcontainer setup

Install Docker and Docker Compose.

Checkout `PMM-2.0` branch which is the main branch for PMM 2.x development.

Run `make` to see a list of targets that can be run on host:
```

> $ make

Reviewer: shoffman@darkhorse ~/Documents/GitHub/pmm $ make
make: *** No targets specified and no makefile found.  Stop.

Reviewer: shoffman@darkhorse ~/Documents/GitHub/pmm $ git status
On branch PMM-2.0
Your branch is up to date with 'origin/PMM-2.0'.

nothing to commit, working tree clean

Author: > shoffman@darkhorse ~/Documents/GitHub/pmm

are you in the correct directory? the path should end with pmm-managed

i think you ran it from the pmm repo (the API repo) which has its Makefile in pmm/api directory  https://github.com/percona/pmm/tree/PMM-2.0/api

Reviewer: Yep, brain fart!

`run` and `run-race` targets replace `/usr/sbin/pmm-managed` and restart pmm-managed with `supervisorctl`. As a result, it will use normal filesystem locations (`/etc/victoriametrics-promscrape.yml`, `/etc/supervisord.d`, etc.) and `pmm-managed` PostgreSQL database. Other locations (inside `testdata`) and `pmm-managed-dev` database are used for unit tests.

# Advanced Setup
## Add instances for monitoring
`make env-up` just starts the PMM server but it doesn't setup anything to be monitored. We can use [pmm-admin](https://github.com/percona/pmm-admin) and [pmm-agent](https://github.com/percona/pmm-agent) to add instances to be monitored to pmm-managed.

> * Clone the pmm-admin [repo](https://github.com/percona/pmm-admin/) and install it by running `make install`.

Reviewer 2: I wonder if we can provide another Docker Compose with with all databases (like in pmm-agent repo) and percona/pmm-client Docker image instead of asking _pmm-managed_ contributors to compile pmm-admin and pmm-agent there.

Alternatively, we should have "How to contribute to PMM 2" guide somewhere as opposed to "How to contribute to pmm-managed", and just link to that document there.

/cc @sanjabonic @BupycHuk Thoughts?

Reviewer: IMHO we need both...since we treat all this things as their own project we need a "how do I contribute to this thing" and an umbrella "how do we contribute to this product"...then each subset can refer you back to the main "These instructions are for contributing to X a subcomponent of PMM...looking to contribute to PMM as a whole?  See here"

* Clone the pmm-agent [repo](https://github.com/percona/pmm-agent).
* Run database instances to be monitored, you can either run your own or use the [`docker-compose.yml`](https://github.com/percona/pmm-agent/blob/master/docker-compose.yml) file provided by pmm-agent to run MySQL, PostgreSQL and MongoDB containers using `make env-up` in the pmm-agent repo (make sure to comment out the `pmm-server` service in the docker-compose file since we are already running pmm-managed in devcontainer).
* Open another shell session and `cd` into the pmm-agent repo, run `make setup-dev` and `make run` to setup and run pmm-agent and connect it to pmm-managed
* In another shell use pmm-admin to add agents to the database instances and start monitoring them using `pmm-admin add mysql --username=root --password=root-password`, `pmm-admin add postgresql --username=pmm-agent --password=pmm-agent-password` and `pmm-admin add mongodb --username=root --password=root-password`.
* Now pmm-managed has started monitoring the databases, login to the web client in your browser to verify. The number of monitored instances will have increased.

> ## Working with STT

Reviewer 2: ```suggestion
## Working with STT

```

The same things with Markdown parsers / renderers

* Add the following environment variables to `docker-compose.yml` to dowload checks from Percona's checks service
```
PERCONA_TEST_SAAS_HOST=check-dev.percona.com:443
PERCONA_TEST_CHECKS_PUBLIC_KEY=RWTg+ZmCCjt7O8eWeAmTLAqW+1ozUbpRSKSwNTmO+exlS5KEIPYWuYdX

> PERCONA_TEST_CHECKS_INTERVAL=10s # set it to any low duration to execute checks quickly

Reviewer 2: Those things should be in docker-compose.yml file, not there.

In general, every thing that is copied (those tokens, pmm-admin / pmm-agent instructions above) will be outdated when the original is updated. Let's avoid copying text where we can.

Author: removed the env vars

```
* Setup the devcontainer using `make env-up` and run your changes inside it by running `make env` and then`make run`.

> * Follow the steps in the previous section to set up instances for monitoring.

Reviewer: ```suggestion
* Follow the steps in the [Add Instances for Monitoring](#add-instances-for-monitoring) to set up instances for monitoring.
```
Will future proof in case a new section ever gets inserted between add and STT

> * Go to the PMM dashboard and enable STT in `PMM Settings -> Advanced Settings`.

Reviewer: ```suggestion
* Go to the PMM dashboard and enable STT in `PMM -> PMM Settings -> Advanced Settings`.
```

* Any failed checks will produce result on the dashboard.

## Working with Integrated Alerting
* Make sure you have instances setup up for monitoring.

> * First we need to setup global communication settings for notification channels. Go to "Communications" tab under "PMM Settings" and set either the SMTP config for email alerts or the Slack webhook URL for slack alerts or both.

Reviewer: ```suggestion
* Enable Integrated Alerting via `PMM -> PMM Settings -> Advanced Settings`.
* Setup global communication settings for notification channels. Go to "Communications" tab under "PMM Settings" and set either the SMTP config for email alerts or the Slack webhook URL for slack alerts or both.
```

Only needed while in tech preview and disabled by default...

* Go to the "Integrated Alerting" dashboard and select the "Notification Channels" tab. Create a notification channel to send the alerts to.
* Now create an alert rule using the built-in alert rule templates and the notification channel.

> * If the alert rule condition satisfies then the alert will be visible on the "Alerts" tab and notification sent to the channel provided that it was configured correctly.

Reviewer 2: All that should be in the user documentation, and there should be just a link with a short explanation. Again, don't copy/duplicate text / instructions

Reviewer: @AlekSi, suggestion is good but take the couple extra seconds to point out the link if possible: https://www.percona.com/doc/percona-monitoring-and-management/2.x/using/alerting.html 
I'd assume that if they weren't here in the first place it's because the author didn't know they existed

Reviewer 2: Neither did I :) But I was sure it exists

Author: updated

## Code Structure
```
.
├── bin - binaries
├── cmd - code for any scripts run by managed
├── data - integrated alerting templates and generated code
├── models - database helpers and types, the database schema can be found in models/database.go file
├── services - contains all the APIs for interacting with services like alertmanager, checks service, victoriametrics, etc
├── testdata - dummy data files used in unit tests
├── utils - utilities
```

# How to make PR
* If the changes require multiple PRs spanning multiple repos make sure to keep the branch names same.
* If the PR requires any API changes then make your changes in `PMM-2.0` branch of the [API repo](https://github.com/percona/pmm) and pull those changes in your pmm-managed branch by mentioning the API changes branch name in the `Gopkg.toml` constraint and running `dep ensure -v -update github.com/percona/pmm`.
* If the PR introduces changes to the IA templates make sure to run `make gen` to udpate the generated code.
* If the PR changes any files named `deps.go` make sure to run `make gen` to generate mock clients.

> * In case a new package is added make sure to udpate `packages.dot` file.

Reviewer 2: This file is generated by main_test.go, no need to update it manually

Before making PR, please run these commands locally:
* `make env TARGET=check-all` to run all checkers and linters.

> * `make env TARGET=test-race` to run tests.

Reviewer: ```suggestion
* `make env TARGET=test-race` to run tests.
* For help, please post on the [PMM 2.x Forums](https://forums.percona.com/c/percona-monitoring-and-management-pmm/percona-monitoring-and-management-pmm-v2/)
```

Author: @shoffman-percona i think our community slack would be a better way for the contributors to reach out to us, should I add that instead or along with the forum link?

Reviewer: My fear of using slack is it sets the impression we are monitoring it...and I don't know who is (I go there on occasion but for now the forums are FAR more active than the community slack

Author: done
